compiler/src/dotty/tools/dotc/transform/init/Summarization.scala

```diff
@@ -213,6 +214,19 @@ object Summarization {
     else summary
   }
 
+  private def effectsOfType(tp: Type, source: Tree)(implicit env: Env): Effects =
+    var summary = Summary.empty
+    val traverser = new TypeTraverser {
+      def traverse(tp: Type): Unit = tp match {
+        case TermRef(_: SingletonType, _) =>
```

Comment:
Why do we restrict to SingletonType here as a prefix of the TermRef?

What could the prefix be other than a SingletonType and why is it OK to skip those cases?

Reply:
In the community build, the project effpi (which use compiler plugin for session protocol checking) generates types like the following:

```
effpi.examples.demo.audit.types.Pay#replyTo
```

The type looks dubious, but the compiler seems to be fine with it. For such a type, our code will simply recur on the prefix `effpi.examples.demo.audit.types.Pay` and look for term selections recursively.

Reply:
I didn't make it clear:  it's a display problem, it's a TermRef where the prefix is a TypeRef. 

If we only have `_: TermRef`, we will need to do more error-handling in `analyze(tp)` to avoid crash: currently it's pretty strict about what types it may get.

Reply:
Oh, sorry, yes, this makes sense now.

Then I just wonder why the restriction is for the prefix to be a `SingletonType` rather than the prefix being something like `TermRef | NoPrefix`. Can a `SingletonType` be more things than that and are those things that we want?

Reply:
Yes, the prefix can also be `ThisType`, which is usually what we care about. If the prefix is `NoPrefix`, the variable is a local variable --- the analysis currently enforces that all local variables are hot.